src/System.ServiceProcess.ServiceController/tests/System.ServiceProcess.ServiceController.TestService/TestService.cs

```diff
@@ -29,66 +34,93 @@ public TestService(string serviceName)
         {
-            WriteLog(nameof(OnContinue));
             base.OnContinue();
+            WriteStreamAsync(PipeMessageByteCode.Continue).Wait();
         }
 
         protected override void OnCustomCommand(int command)
         {
-            WriteLog(nameof(OnCustomCommand) + " command=" + command);
             base.OnCustomCommand(command);
+            WriteStreamAsync(PipeMessageByteCode.OnCustomCommand, command).Wait();
         }
 
         protected override void OnPause()
         {
-            WriteLog(nameof(OnPause));
             base.OnPause();
+            WriteStreamAsync(PipeMessageByteCode.Pause).Wait();
         }
 
         protected override void OnSessionChange(SessionChangeDescription changeDescription)
         {
-            WriteLog(nameof(OnSessionChange) + " change=" + changeDescription.ToString());
             base.OnSessionChange(changeDescription);
         }
 
         protected override bool OnPowerEvent(PowerBroadcastStatus powerStatus)
         {
-            WriteLog(nameof(OnPowerEvent) + " status=" + powerStatus.ToString());
             return base.OnPowerEvent(powerStatus);
         }
 
         protected override void OnShutdown()
         {
-            WriteLog(nameof(OnShutdown));
             base.OnShutdown();
         }
 
         protected override void OnStart(string[] args)
         {
-            File.Delete(GetLogPath(ServiceName));
-
-            WriteLog(nameof(OnStart) + " args=" + string.Join(",", args));
             base.OnStart(args);
+            if (args.Length == 4 && args[0] == "StartWithArguments")
+            {
+                Debug.Assert(args[1] == "a");
+                Debug.Assert(args[2] == "b");
+                Debug.Assert(args[3] == "c");
+                WriteStreamAsync(PipeMessageByteCode.Start).Wait();
+            }
         }
 
         protected override void OnStop()
         {
-            WriteLog(nameof(OnStop));
             base.OnStop();
+            WriteStreamAsync(PipeMessageByteCode.Stop).Wait();
+        }
+
+        private async Task WriteStreamAsync(PipeMessageByteCode code, int command = 0)
+        {
+            const int writeTimeout = 60000;
+            Task writeCompleted;
+            if (_waitClientConnect.IsCompleted)
+            {
+                if (code == PipeMessageByteCode.OnCustomCommand)
+                {
+                    writeCompleted = _serverStream.WriteAsync(new byte[] { (byte)command }, 0, 1);
+                    await writeCompleted.TimeoutAfter(writeTimeout).ConfigureAwait(false);
+                }
+                else
+                {
+                    writeCompleted = _serverStream.WriteAsync(new byte[] { (byte)code }, 0, 1);
+                    await writeCompleted.TimeoutAfter(writeTimeout).ConfigureAwait(false);
+                }
+            }
+            else
+            {
+                throw new TimeoutException($"Client didn't connect to the pipe after 30 seconds");
```

Comment:
nit: you don't know for a fact if it was 30 seconds, you just know that the server is still waiting

Reply:
okay I will change it and merge